src/main/java/com/github/firmwehr/gentle/firm/optimization/GlobalValueNumbering.java

```diff
@@ -377,15 +385,15 @@ private void onEnter(Block block) {
 				}
 			}
 
-			// local block stabilized, update available expression
-			availableExpressions.putAll(currentAvailable);
-
 			// extract surviving phi nodes, we might need to update them later
-			for (var node : availableExpressions.values()) {
+			for (var node : currentAvailable.values()) {
 				if (node instanceof Phi phi) {
 					phis.add(phi);
 				}
 			}
+
+			// local block stabilized, update available expression
+			availableExpressions.putAll(currentAvailable);
```

Comment:
@I-Al-Istannen This should remove repeated insert ops on the `phis`-Set and thous improve performance.